paths_cli/compiling/core.py

```diff
@@ -0,0 +1,300 @@
+import yaml
+
+from collections import namedtuple
+
+import logging
+
+from .errors import InputError
+from paths_cli.utils import import_thing
+from paths_cli.plugin_management import OPSPlugin
+
+def listify(obj):
+    listified = False
+    if not isinstance(obj, list):
+        obj = [obj]
+        listified = True
+    return obj, listified
+
+def unlistify(obj, listified):
+    if listified:
+        assert len(obj) == 1
+        obj = obj[0]
+    return obj
+
+REQUIRED_PARAMETER = object()
+
+class Parameter:
+    SCHEMA = "http://openpathsampling.org/schemas/sim-setup/draft01.json"
+    def __init__(self, name, loader, *, json_type=None, description=None,
+                 default=REQUIRED_PARAMETER, aliases=None):
+        if isinstance(json_type, str):
+            try:
+                json_type = json.loads(json_type)
+            except json.decoder.JSONDecodeError:
+                # TODO: maybe log this in case it represents an issue?
+                pass
+
+        self.name = name
+        self.loader = loader
+        self._json_type = json_type
+        self._description = description
+        self.default = default
+
+    @property
+    def required(self):
+        return self.default is REQUIRED_PARAMETER
+
+    @property
+    def json_type(self):
+        return self._get_from_loader(self.loader, 'json_type',
+                                     self._json_type)
+
+    @property
+    def description(self):
+        return self._get_from_loader(self.loader, 'description',
+                                     self._description)
+
+    @staticmethod
+    def _get_from_loader(loader, attr_name, attr):
+        if attr is None:
+            try:
+                attr = getattr(loader, attr_name)
+            except AttributeError:
+                pass
+        return attr
+
+    def __call__(self, *args, **kwargs):
+        # check correct call signature here?
+        return self.loader(*args, **kwargs)
+
+    def to_json_schema(self, schema_context=None):
+        dct = {
+            'type': self.json_type,
+            'description': self.description,
+        }
+        return self.name, dct
+
+
+class Builder:
+    """Builder is a wrapper class to simplify writing builder functions.
+
+    When the compiled parameters dictionary matches the kwargs for your
+    class, you can create a valid delayed builder function with
+
+    .. code::
+
+        builder = Builder('import_path.for_the.ClassToBuild')
+
+    Additionally, this class provides hooks for functions that run before or
+    after the main builder function. This allows many objects to be built by
+    implementing simple functions and hooking themn together with Builder.
+
+    Parameters
+    ----------
+    builder : Union[Callable, str]
+        primary callable to build an object, or string representing the
+        fully-qualified path to a callable
+    remapper : Callable[[Dict], Dict], optional
+        callable to remap the the mapping of ???
```

Comment:
missing text? or is it literal "???"

Reply:
Was definitely missing text. I often use "???" as a reminder to fill this in later -- obviously missed that one! Fixed now.